### PR TITLE
[CDAP-18701] Fix program state where it intermittently transitions to…

### PR DESCRIPTION
**Pull request by CuriousVini**
… STOPPED state

Tested with 17-80 consecutive runs. Before issue used to get reproduced multiple times in 10 runs or so.

### CODE REVIEW

File: cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillController.java [hunk 1]
```diff
@@ -526,36 +536,24 @@ private void deletePVCs() {
   private CompletionStage<String> deleteJobAndGetCompletionStatus() {
     CompletableFuture<String> resultFuture = new CompletableFuture<>();
     String name = meta.getName();
-    try {
-      V1Job v1Job = batchV1Api.readNamespacedJob(name, kubeNamespace, null, null, null);
-      V1JobStatus status = v1Job.getStatus();
-
+    if (jobStatus == null) {
+      isStopped = true;
+      resultFuture.complete(name);
```
Comment (masoud-io) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillController.java:556]: If we return here, we don't need `else` in the next line.

File: cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillController.java [hunk 2]
```diff
@@ -526,36 +536,24 @@ private void deletePVCs() {
...
+    } else if (jobStatus.getFailed() != null) {
       // If job has failed, mark future as failed. Else mark it as succeeded.
-      if (status != null && status.getFailed() != null) {
-        resultFuture.completeExceptionally(new RuntimeException(String.format("Job %s has failed status.", name)));
-      } else {
-        resultFuture.complete(name);
-      }
+      resultFuture.completeExceptionally(new RuntimeException(String.format("Job %s has failed status.", name)));
```
Comment (masoud-io) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillController.java]: `has a failed status`

File: cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillController.java [hunk 3]
```diff
@@ -69,13 +69,15 @@
...
   private final RunId runId;
   private final CompletableFuture<KubeTwillController> completion;
-  private boolean isStopped;
   private final DiscoveryServiceClient discoveryServiceClient;
   private final ApiClient apiClient;
   private final BatchV1Api batchV1Api;
   private final Type resourceType;
   private final V1ObjectMeta meta;
 
+  private volatile boolean isStopped;
```
Comment (masoud-io) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillController.java:79]: Looks like we are not using this field. Do we need it?

**Review by masoud-io — COMMENTED** [review]
This fix should be merged into `develop` branch first. Right?

File: cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillController.java (hunk 3 above)
Comment (CuriousVini) [reply cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillController.java:79]: Its still being used in line 540

File: cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillController.java (hunk 1 above)
Comment (CuriousVini) [reply cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillController.java:556]: We still need to attempt to delete the job.

**Comment by CuriousVini** [comment]
> This fix should be merged into `develop` branch first. Right?

Yea, I will open it once this is merged since its already reviewed

**Review by masoud-io — APPROVED** [review]

**Pull request merged**